src/components/uui-pagination/uui-pagination.element.ts

```diff
@@ -0,0 +1,306 @@
+      this.current
+    }.`;
+  }
+
+  private _containerWidth = 0;
+
+  private calculateRange = () => {
+    this._containerWidth = this.offsetWidth;
+
+    // get all the buttons with .nav-button class and sum up thier widths
+    const navButtonsWidth = Array.from(this.navButtons).reduce(
+      (totalWidth, button) => {
+        return totalWidth + button.getBoundingClientRect().width;
+      },
+      0
+    );
+
+    // substract width of navbuttons from the pagination container
+    const rangeBaseWidth = this._containerWidth - navButtonsWidth;
+
+    // divide remaining width by max-width of page button (when it has 3 digits), then divide by 2 to get the range.
+    // Range is number of buttons visible on either "side" of current pag button. So, if range === 5 we shall see 11 buttons in total - 5 before the curent page and 5 after. This is why we divide by 2.
+    const range = rangeBaseWidth / PAGE_BUTTON_MAX_WIDTH / 2;
+    this.range = Math.floor(range);
+  };
+
+  private _generateArrayOfNumbers(start: number, stop: number) {
+    return Array.from({ length: stop - start + 1 }, (_, i) => start + i);
+  }
+
+  private _generateVisiblePages(current: number) {
+    const start =
+      current < this._range
+        ? 1
+        : current < this.total - this._range
+        ? current - this._range
+        : this.total - this._range * 2;
+
+    const stop =
+      current <= this._range
+        ? this._range * 2 + 1
+        : current < this.total - this._range
+        ? current + this._range
+        : this.total;
+
+    const pages = this._generateArrayOfNumbers(
+      this._valueLimit(start, 1, this.total),
+      this._valueLimit(stop, 1, this.total)
+    );
+
+    return pages;
+  }
+
+  private _valueLimit(val: number, min: number, max: number) {
+    return Math.min(Math.max(val, min), max);
+  }
+
+  @queryAll('uui-button.nav-button')
+  navButtons!: Array<UUIButtonElement>;
+
+  @property()
+  label = '';
+
+  // TODO: Handle localization
+  @property({ reflect: true, attribute: 'aria-label' })
+  ariaLabel = '';
+
+  @property({ type: Number })
+  total = 1;
+
+  protected _range = 0;
+  @property({ type: Number })
+  get range() {
+    return this._range;
+  }
+
+  set range(newValue: number) {
+    const oldValue = this._range;
+    this._range = newValue <= 0 ? 1 : newValue;
+    this.visiblePages = this._generateVisiblePages(this.current);
+    this.requestUpdate('range', oldValue);
+  }
+
+  @state()
+  visiblePages: number[] = [];
+
+  protected _current = 1;
+  @property({ type: Number, reflect: true })
+  get current() {
+    return this._current;
+  }
+
+  set current(newValue: number) {
+    const oldValue = this._current;
+    this._current = this._valueLimit(newValue, 1, this.total);
+    this.visiblePages = this._generateVisiblePages(this._current);
+    this.requestUpdate('current', oldValue);
+  }
+
+  protected goToNextPage() {
```

Comment:
These methods are made 'protected' so they cant be called from the outside, we don't want outsiders to call methods that fires Events, as the Events are equal to interaction.